pallets/msa/src/benchmarking.rs

```diff
@@ -157,16 +157,20 @@ benchmarks! {
 	}: _(RawOrigin::Signed(caller_and_delegator_public_key), new_public_key)
 
 	retire_msa {
-
 		let caller: T::AccountId = whitelisted_caller();
-
-		// Create a MSA account
-		assert_ok!(Msa::<T>::create(RawOrigin::Signed(caller.clone()).into()));
-		let msa_id = Msa::<T>::ensure_valid_msa_key(&caller).unwrap();
-
-		assert_eq!(Msa::<T>::is_registered_provider(msa_id),false);
-
-	}: _(RawOrigin::Signed(caller))
+		assert_ok!(Msa::<T>::add_key(ProviderId(1).into(), &caller.clone(), EMPTY_FUNCTION));
```

Comment:
I think you will need to have something like
```
let n in 2 .. MAX_NUMBER_OF_PROVIDERS_PER_DELEGATOR;
for j in 2..=n {
    assert_ok!(Msa::<T>::add_provider(ProviderId(j.into()), DelegatorId(1), vec![1, 2]));
}
```

Reply:
this will allow the benchmark function to find the multiplier correctly instead of using the maximum as a constant

Reply:
We were talking about something like this but we were not certain about it. Can you explain what you mean?
Why not use a constant since it is the worst-case scenario?

>You can then use the results of the benchmarking tests to establish an approximate worst case weight to represent the resources required to execute each function call and each code path

https://docs.substrate.io/test/benchmark/

Reply:
Because this worst case is not accurate. Calculated weights are substantially bigger than the correct one and it is unnecessarily consuming a huge chunk of block size.

Reply:
I think that weight gets refunded after the extrinsic runs.

Reply:
I don't think it is.

Reply:
https://github.com/paritytech/substrate/blob/master/frame/system/src/extensions/check_weight.rs#L246

Reply:
I don't think that is related for this case

Reply:
`calc_unspent` Calculate how much (if any) weight was not used by the `Dispatchable`. But we are not specifying how much actually we used in our return

Reply:
You are right maybe we should be returning in this case. Also, perhaps if we knew how many delegations needed to be deleted it would make for better estimates.

Reply:
I think this is a good argument to keep count of how many delegation one has.

Reply:
I am thinking. Maybe we ask the user to provide how many delegations they have + a signed extension to validate the provided count vs the actual count. We drop the transaction if it does not match.

Reply:
What would be the benefit of doing that?

Reply:
This could potentially block other transactions from making it into the block. 128 writes a lot.